src/middleware/cors/service.rs

```diff
@@ -1,41 +1,211 @@
-use super::Config;
+use super::{Config, CorsResource};
 
-use futures::{Poll};
-use http;
+use futures::{Async, Future, Poll};
+use http::{self, HeaderMap, StatusCode};
+use http::{Request, Response};
 use tower_service::Service;
 
 use std::sync::Arc;
 
+#[derive(Debug)]
 pub struct CorsService<S> {
     inner: S,
     config: Arc<Config>,
 }
 
 impl<S> CorsService<S> {
     pub(super) fn new(inner: S, config: Arc<Config>) -> CorsService<S> {
-        CorsService {
-            inner,
-            config,
-        }
+        CorsService { inner, config }
     }
 }
 
-use util::http::HttpService;
-use http::{Request, Response};
-
 impl<S> Service for CorsService<S>
-where S: HttpService,
+where
+    S: ::util::http::HttpService,
 {
     type Request = Request<S::RequestBody>;
-    type Response = Response<S::ResponseBody>;
+    type Response = Response<Option<S::ResponseBody>>;
     type Error = S::Error;
-    type Future = S::Future;
+    type Future = CorsFuture<S::Future>;
 
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
         self.inner.poll_ready()
     }
 
     fn call(&mut self, request: Self::Request) -> Self::Future {
-        self.inner.call(request)
+        match self.config.process_request(&request) {
+            Ok(CorsResource::Preflight(headers)) => CorsFuture::Handled(Some(headers)),
+            Ok(CorsResource::Simple(headers)) => {
+                CorsFuture::Simple(self.inner.call(request), Some(headers))
+            }
+            Err(e) => {
+                eprintln!("{:?}", e);
```

Comment:
Before merging, likely want to remove this println. As to the question from the previous PR about what exactly to do, here's some options:

- The [CORS spec](https://fetch.spec.whatwg.org/#http-cors-protocol) mentions that if the `Access-Control-Allow-Origin` header isn't set, then the browser will prevent the JS context from accessing the response. So, one option is to just call the inner service like normal, and simply not add the CORS headers, and use the `log` crate to explain what happened.
- The spec suggests that servers *can* return a 403 if the origin or related headers are wrong. This looks like what is currently done. In this case, the only change would be to replace the println with a `debug!` or something.
- For maximum control, the error could be returned. That would change the `Service::Error` type to be some sort of `Either<Inner, Cors>`. Then, something higher up the chain can determine exactly what to do if CORS isn't allowed.

Reply:
I went with the `debug` for now, as it's accessible to a developer but doesn't require a bunch of extra work to get to.